### PR TITLE
HPCC-26717 hThor file costs (all except index reads)

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: ecl/hthor/hthor.cpp [hunk 1]
```diff
@@ -752,11 +757,24 @@ void CHThorDiskWriteActivity::publish()
         Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(desc);
         if(file->getModificationTime(modifiedTime))
             file->setAccessedTime(modifiedTime);
+        StringBuffer clusterName;
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp:760]: minor: move clusterName inside the if()

File: ecl/hthor/hthor.cpp [hunk 2]
```diff
@@ -1211,6 +1229,7 @@ void CHThorIndexWriteActivity::execute()
         builder->finish(metadata, &fileCrc);
         duplicateKeyCount = builder->getDuplicateCount();
         cummulativeDuplicateKeyCount += duplicateKeyCount;
+        numDiskWrites = io->getStatistic(StNumDiskWrites);
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp:1232]: Slight question about if this was executed in a child query, but that is so unusual that it doesn't matter...

File: ecl/hthor/hthor.cpp [hunk 3]
```diff
@@ -8414,6 +8437,32 @@ unsigned __int64 CHThorDiskReadBaseActivity::getLocalFilePosition(const void * r
...
+                if (superfile)
+                {
+                    unsigned subfile, lnum;
+                    if (superfile->mapSubPart(previousPartNum, subfile, lnum))
+                    {
+                        IDistributedSuperFile *super = dFile->querySuperFile();
+                        dFile = &(super->querySubFile(subfile, true));
+                    }
+                }
+                dFile->addAttrValue("@numDiskReads", curDiskReads);
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp:8458]: @jakesmith is there any issue with updating these properties?  Is the dali entry locked, and does it happen straight away.  What implications does this have once esp provides the meta information!

File: ecl/hthor/hthor.cpp [hunk 4]
```diff
@@ -8446,8 +8496,7 @@ bool CHThorDiskReadBaseActivity::openNext()
             unsigned numCopies = curPart?curPart->numCopies():ldFile->numPartCopies(partNum);
             //MORE: Order of copies should be optimized at this point....
             StringBuffer file, filelist;
-            closepart();
-            if (dfsParts && superfile && curPart)
+            if (dfsParts && superfile && curPart && !subfileLogicalFilenames.empty())
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp:8499]: Has the closepart() moved to the correct place?  Would it also work to call closepart() after the while loop?  Asking because there is another call to closepart() within the loop code, and not sure exactly why the change was needed (to make sure stats were updated?)

**Review by ghalliday — COMMENTED** [review]
@shamser a few, fairly minor comments.

File: common/workunit/workunit.cpp [hunk 5]
```diff
@@ -2620,6 +2620,40 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
...
+        case SSTfunction:
+        case SSTchannel:
+        case SSTglobal:
+            continue;
+        default:
+            throwUnexpected();
+        }
+    }
+    return totalCost;
+}
```
[the remark this answers is not in the document]
Comment (shamser) [reply common/workunit/workunit.cpp:2655]: I'll update to use HPCC-26788

File: common/workunit/workunit.cpp [hunk 6]
```diff
@@ -2620,6 +2620,40 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
...
+
```
[the remark this answers is not in the document]
Comment (shamser) [reply common/workunit/workunit.cpp:2656]: All the places that call aggregateDiskAccessCost(const IConstWorkUnit * wu, const char *scope) doesn't have access to IStatisticCollection.  e.g. EclAgent::doProcess(), EclAgentWorkflowMachine::noteTiming() and EclGraph::execute().  I'm not sure if I can get the required IStatisticCollection, from a call to a member function of IConstWorkUnit

File: common/workunit/workunit.hpp [hunk 7]
```diff
@@ -1708,6 +1708,7 @@ extern WORKUNIT_API void updateWorkunitTimings(IWorkUnit * wu, ITimeReporter *ti
 extern WORKUNIT_API void updateWorkunitTimings(IWorkUnit * wu, StatisticScopeType scopeType, StatisticKind kind, ITimeReporter *timer);
 extern WORKUNIT_API void aggregateStatistic(StatsAggregation & result, IConstWorkUnit * wu, const WuScopeFilter & filter, StatisticKind search);
 extern WORKUNIT_API cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope=nullptr, bool excludehThor=false);
+extern WORKUNIT_API cost_type aggregateDiskAccessCost(IStatisticCollection * statsCollection);
```
[the remark this answers is not in the document]
Comment (shamser) [reply common/workunit/workunit.hpp:1711]: Removed.  No longer needed. Using AggregateStatistic() instead.

File: common/workunit/workunit.cpp [hunk 8]
```diff
@@ -2620,6 +2620,40 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
...
+    cost_type totalCost = 0;
+    Owned<IStatisticCollectionIterator> iter = &(statsCollection->getScopes(nullptr, false));
+    ForEach(*iter)
+    {
+        IStatisticCollection & cur = iter->query();
+        switch (cur.queryScopeType())
+        {
+        case SSTactivity:
+            totalCost += cur.queryStatistic(StCostFileAccess);
+            break;
```
[the remark this answers is not in the document]
Comment (shamser) [reply common/workunit/workunit.cpp:2634]: The new aggregateStatistic() is including the child activities costs.

File: ecl/hthor/hthor.cpp (hunk 4 above)
Comment (shamser) [reply ecl/hthor/hthor.cpp:8501]: I realize that it's not actually reading the file within that loop.  I've moved closepart back to the original location.

**Review by ghalliday — APPROVED** [review]
I can't see any problems.  All looks good afaics.

File: ecl/hthor/hthor.cpp (hunk 3 above)
Comment (jakesmith) [reply ecl/hthor/hthor.cpp:8460]: addAttrValue automatically locks the file attr properties, for an atomic update, and the update is pushed immediately.
There will need to be an implemtation of this that does the same atomic update when it's via a esp service.

Comment (ghalliday) [reply ecl/hthor/hthor.cpp:8460]: Yes, I added a subtask/comment to the jira to make sure it was covered.
